### PR TITLE
Fixes #25885 - Checksum-type updates on smart proxy

### PR DESCRIPTION
**Pull request by sjha4**
### Description
*****
Pulp calculates and stores the checksum_type of repo based on the metadata and packages and stores it on scratchpad. That should also be the source of truth for checksum-type in capsule instead of setting it directly.
Therefore, we can ignore checksum_type when syncing repos between capsules. Because pulp will invariably calculate the checksum_type. Our overriding of that checksum_type can lead to inconsistency as highlighted by the issue in question.
****
The second part to this is the ability to set checksum_type on repos with on_demand download policy. In case of immediate download policy, the distributor can actually publish with the checksum_type specified. However, changing checksum type on repos with on_demand can actually lead to errors when publishing. To that end, we should disallow users from setting checksum_type on yum repos with on_demand download policy. 
****
#### Part 1: 
1. Create a product and repository with sha256
2. Add this a lifecycle environment related to a capsule and sync it
3. Use pulp-admin or check the repodata created on the capsule and it will be using sha256
4. Modify the repository to use sha1 and create a new version for the content view
5. Sync and check the repodata. On the Satellite server, repodata now will have sha1 however the capsule still has sha256

**Modified behavior:**
We do not set the checksum_type on the repo in the capsule anymore and let pulp calculate it and put it on the scratchpad instead.

#### Part 2:
1. Create a yum repo with on_demand download policy
2. Set the checksum_type to any value other than Default (nil)

**Modified behavior:**
Validation error thrown when creating/updating yum repos with on_demand download policy with checksum_type set.
****
Added config to re-run the upgrade task to clear checksum_type on on_demand yum repos which was also part of 3.8 upgrade.

### CODE REVIEW

**Comment by jturel** [comment]
I can't bring up a proxy box right now, will check when it's fixed

File: app/services/katello/pulp/repository/yum.rb [hunk 1]
```diff
@@ -34,8 +34,7 @@ def generate_distributors
           options = {
             protected: true,
             id: yum_dist_id,
-            auto_publish: true,
-            checksum_type: repo.saved_checksum_type || root.checksum_type
+            auto_publish: true
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository/yum.rb]: we still want checksum_type to be set if its a pulp_master, right?

Comment (sjha4) [reply app/services/katello/pulp/repository/yum.rb]: Updated to set checksum_type for pulp_master and rely on scratch pad on capsule.

**Comment by sjha4** [comment]
Updated cassettes to resolve test issues..

File: app/models/katello/root_repository.rb [hunk 2]
```diff
@@ -129,6 +130,12 @@ def ensure_no_download_policy
       end
     end
 
+    def ensure_no_checksum_on_demand
+      if checksum_type.present? && [::Runcible::Models::YumImporter::DOWNLOAD_ON_DEMAND].include?(download_policy)
```
Comment (jturel) [inline app/models/katello/root_repository.rb]: Should this just be `::Runcible::Models::YumImporter::DOWNLOAD_ON_DEMAND == download_policy` ?

**Comment by jturel** [comment]
Works as described! Left one note. Is the scratchpad an arbiratry key => value store for us to use?

**Pull request merged**